nativelib/src/main/scala/scala/scalanative/runtime/Arrays.scala.gyb

```diff
@@ -33,88 +38,230 @@ sealed abstract class Array[T]
+   */   
+  def alloc(length: Int, arrinfo:  Ptr[Type], sizeofElement: CSize): Ptr[_] = {
+    val arrsize = sizeof[ArrayHeader] + sizeofElement * length
+    val arr = runtime.alloc(arrinfo, arrsize)
     // set the length
-    !(arr.cast[Ptr[Byte]] + sizeof[Ptr[_]]).cast[Ptr[Int]] = length
-    arr.cast[ObjectArray]
+    !(arr.cast[Ptr[Byte]] + sizeof[Ptr[_]]).cast[Ptr[Int]] = length    
+    arr        
+  } 
+}
+
+/*
+ * 
+ */
+ 
+abstract class PrimitiveArray[T] extends Array[T] {
+  /** called by subclass::clone.  Allocate memory using GC_malloc_atomic and copy header and data */
+  override protected def helperClone(sizeofElement: CSize): Ptr[_] = {
+    val arrsize = sizeof[ArrayHeader] + sizeofElement * length
+    val arr = runtime.allocAtomic(arrsize)
+    `llvm.memcpy.p0i8.p0i8.i64`(arr.cast[Ptr[Byte]], this.cast[Ptr[Byte]], arrsize, 1, false)
+    arr
+  }  
+}
+
+/*
+ * 
+ */
+
+object PrimitiveArray {
+  /*
+   * Primitive arrays don't contain pointers 
+   */   
+  def alloc(length: Int, arrinfo:  Ptr[Type], sizeofElement: CSize): Ptr[_] = {
+    val arrsize = sizeof[ArrayHeader] + sizeofElement * length
+    val arr = runtime.allocAtomic(arrinfo, arrsize)
+    // set the length
+    !(arr.cast[Ptr[Byte]] + sizeof[Ptr[_]]).cast[Ptr[Int]] = length    
+    arr        
   }
 }
 
+/*
+ *
+ */
+final class ObjectArray private () extends Array[Object] {
+  def apply(i: Int): Object = ! (pointerAt(sizeof[Object], i).cast[Ptr[Object]])
+
+  def update(i: Int, value: Object): Unit = ! (pointerAt(sizeof[Object], i).cast[Ptr[Object]]) = value
+
+  protected override def clone(): ObjectArray = helperClone (sizeof[Object]).cast[ObjectArray]  
+}
+
+object ObjectArray {
+  def alloc(length: Int): ObjectArray = Array.alloc(length, typeof[ObjectArray], sizeof[Object]).cast[ObjectArray]
+}
+
 % for T in primitiveTypes:
 
-final class ${T}Array private () extends Array[${T}] {
-  def apply(i: Int): ${T} =
-    if (i < 0 || i >= length)
-      throw new IndexOutOfBoundsException(i.toString)
-    else {
-      val headptr = (this.cast[Ptr[Byte]] + sizeof[ArrayHeader]).cast[Ptr[${T}]]
-      headptr(i)
-    }
+final class ${T}Array private () extends PrimitiveArray[${T}] {
+  def apply(i: Int): ${T} = ! (pointerAt(sizeof[${T}], i).cast[Ptr[${T}]])
 
-  def update(i: Int, value: ${T}): Unit =
-    if (i < 0 || i >= length)
-      throw new IndexOutOfBoundsException(i.toString)
-    else {
-      val headptr = (this.cast[Ptr[Byte]] + sizeof[ArrayHeader]).cast[Ptr[${T}]]
-      headptr(i) = value
-    }
+  def update(i: Int, value: ${T}): Unit = ! (pointerAt(sizeof[${T}], i).cast[Ptr[${T}]]) = value
 
-  protected override def clone(): ${T}Array = {
-    val newarr = ${T}Array.alloc(length)
-    ${T}Array.copy(this, 0, newarr, 0, length)
-    newarr
-  }
+  protected override def clone(): ${T}Array = helperClone (sizeof[${T}]).cast[${T}Array]
+  
+  protected def copyTo (fromPos: Int, toPos: Int): Unit = update (toPos, apply(fromPos)) 
```

Comment:
Please don't put helper methods, which are not an implementation of Array's public API. All the helper methods should stay in the companions.